helix-term/src/commands.rs

```diff
@@ -6009,6 +6048,54 @@ fn replay_macro(cx: &mut Context) {
     }));
 }
 
+fn replay_macro_picker(cx: &mut Context) {
+    let register = cx.register.unwrap_or('@');
+
+    if cx.editor.macro_replaying.contains(&register) {
+        cx.editor.set_error(format!(
+            "Cannot replay from register [{}] because already replaying from same register",
+            register
+        ));
+        return;
+    }
+    cx.editor.autoinfo = Some(Info::from_selected_register(&cx.editor.registers, register));
+    cx.on_next_key(move |cx, event| {
+        cx.editor.autoinfo = None;
+
+        let Some(ch) = event.char() else {
+            return;
+        };
+        let Ok(index) = usize::from_str_radix(&ch.to_string(), 16) else {
+            log::debug!("During replay macro picker, {ch:?} was not parsed as a number");
+            return;
+        };
+        match parse_keys_from_register_index(cx, register, index) {
+            Ok(keys) => {
+                replay_macro_impl(cx, register, keys);
+            }
+            Err(err) => {
+                cx.editor.set_error(format!(
+                    "Register [{register}] @ index {index}: invalid macro: {err}"
+                ));
+            }
+        }
+    })
+}
+
+fn parse_keys_from_register_index(
```

Comment:
I needed to break this out because of a conflict between mutable and immutable borrows of Context and Editor.